### PR TITLE
Update JEP-12049 implementation discussion

### PR DESCRIPTION
**Pull request by jakevdp**
Updated plan based on our discussion this morning.

### CODE REVIEW

File: docs/jep/12049-type-annotations.md [hunk 1]
```diff
@@ -284,24 +284,27 @@ Here `jnp.ndarray` could be an alias for `jax.Array`.
...
-We could appease OOP pedants by instead making `Tracer` and `Array` derive from a common `ArrayBase` base class:
+We could make the class hierarchy more sensible by making `Tracer` and the class for
+on-device arrays inherit from a common base class. So, for example:
 
-- `jax.Array` is the actual type of on-device arrays
-- `ArrayBase` is the object used for array annotations
-- `ArrayBase` is also the object used for instance checks
+- `jax.Array` is a base class for `Tracer` as well as the actual type of on-device arrays,
+  which might be `jax._src.ArrayImpl` or similar.
+- `jax.Array` is the object used for array annotations
```
Comment (yashk2810) [inline docs/jep/12049-type-annotations.md:292]: nit: type annotations?

**Review by yashk2810 — APPROVED** [review]

**Comment by jakevdp** [comment]
RTD failure is this: https://github.com/google/jax/pull/12422

File: docs/jep/12049-type-annotations.md (hunk 1 above)
Comment (jakevdp) [reply docs/jep/12049-type-annotations.md]: Done (here and elsewhere)

File: docs/jep/12049-type-annotations.md [hunk 2]
```diff
@@ -277,31 +277,34 @@ Here, `jax.numpy.ndarray` would become a simple alias `jax.Array` for backward c
...
 
-- `jax.Array` is the actual type of on-device arrays
-- `ArrayBase` is the object used for array annotations
-- `ArrayBase` is also the object used for instance checks
+- `jax.Array` is a base class for `Tracer` as well as the actual type of on-device arrays,
+  which might be `jax._src.ArrayImpl` or similar.
+- `jax.Array` is the object used for array type annotations
+- `jax.Array` is also the object used for instance checks
 
 Here `jnp.ndarray` would be an alias for `ArrayBase`.
```
Comment (hawkinsp) [inline docs/jep/12049-type-annotations.md:295]: `s/ArrayBase/Array` here?

File: docs/jep/12049-type-annotations.md [hunk 3]
```diff
@@ -277,31 +277,34 @@ Here, `jax.numpy.ndarray` would become a simple alias `jax.Array` for backward c
...
+- From a user perspective, the unified approaches (options 2 and 3) are arguably best, because they remove the cognitive overhead involved in remembering which objects to use for instance checks or annotations: `jax.Array` is all you need to know.
+- However, both options 2 and 3 introduce some strange and/or confusing behavior. Option 2 depends on potentially confusing overrides of instance checks, which are [not well supported](https://github.com/pybind/pybind11/issues/2696) for classes defined in pybind11. Option 3 requires `Tracer`to be a subclass array. This breaks the inheritance model, because it would require `Tracer` objects to carry all the baggage of `Array` objects (data buffers, sharding, devices, etc.)
```
Comment (hawkinsp) [inline docs/jep/12049-type-annotations.md:304]: missing space "`Tracer`to be"

**Review by hawkinsp — APPROVED** [review]

File: docs/jep/12049-type-annotations.md (hunk 2 above)
Comment (jakevdp) [reply docs/jep/12049-type-annotations.md:295]: Thanks!

File: docs/jep/12049-type-annotations.md (hunk 3 above)
Comment (jakevdp) [reply docs/jep/12049-type-annotations.md:304]: Thanks!

**Pull request merged**